src/main/java/org/folio/circulation/services/RequestAnonymizationService.java

```diff
@@ -0,0 +1,134 @@
+import org.folio.circulation.domain.RequestStatus;
+import org.folio.circulation.support.Clients;
+import org.folio.circulation.support.RecordNotFoundFailure;
+import org.folio.circulation.support.ValidationErrorFailure;
+import org.folio.circulation.support.http.server.ValidationError;
+import org.folio.circulation.support.results.Result;
+import org.folio.circulation.infrastructure.storage.inventory.ItemRepository;
+import org.folio.circulation.infrastructure.storage.loans.LoanRepository;
+import org.folio.circulation.infrastructure.storage.users.UserRepository;
+import org.folio.circulation.infrastructure.storage.requests.RequestRepository;
+
+import io.vertx.core.json.JsonObject;
+
+public class RequestAnonymizationService {
+  private static final Set<RequestStatus> ALLOWED_STATUSES = EnumSet.of(
+    RequestStatus.CLOSED_FILLED,
+    RequestStatus.CLOSED_CANCELLED,
+    RequestStatus.CLOSED_PICKUP_EXPIRED,
+    RequestStatus.CLOSED_UNFILLED
+  );
+
+  private static final Set<RequestStatus> OPEN_STATUSES = EnumSet.of(
+    RequestStatus.OPEN_AWAITING_PICKUP,
+    RequestStatus.OPEN_IN_TRANSIT,
+    RequestStatus.OPEN_NOT_YET_FILLED
+  );
+
+  private final RequestRepository requestRepository;
+  private final EventPublisher eventPublisher;
+
+  public RequestAnonymizationService(Clients clients, EventPublisher eventPublisher) {
+    ItemRepository itemRepository = new ItemRepository(clients);
+    UserRepository userRepository = new UserRepository(clients);
+    LoanRepository loanRepository = new LoanRepository(clients, itemRepository, userRepository);
+
+    this.requestRepository = RequestRepository.using(
+      clients,
+      itemRepository,
+      userRepository,
+      loanRepository
+    );
+
+    this.eventPublisher = eventPublisher;
+  }
+
+  public RequestAnonymizationService(RequestRepository requestRepository,
+      EventPublisher eventPublisher) {
+
+    this.requestRepository = requestRepository;
+    this.eventPublisher = eventPublisher;
+  }
+
+  public CompletableFuture<Result<String>> anonymizeSingle(String requestId) {
+    final UUID id;
+    try {
+      id = UUID.fromString(requestId);
+    } catch (IllegalArgumentException e) {
+      return CompletableFuture.completedFuture(
+        failed(new ValidationErrorFailure(new ValidationError("invalidRequestId", "requestId", requestId)))
+      );
+    }
+
+    return fetchRequest(id)
+      .thenApply(flatMapResult(req -> validateStatus(req, requestId)))
+      .thenApply(flatMapResult(this::scrubPii))
+      .thenCompose(r -> r.after(requestRepository::update))
+      .thenCompose(r -> r.after(this::publishLog))
+      .thenApply(mapResult(updated -> requestId));
+  }
+
+  private CompletableFuture<Result<Request>> fetchRequest(UUID id) {
+    return requestRepository.getById(id.toString())
+      .thenApply(res -> res.failWhen(
+        r -> succeeded(r == null),
+        r -> new RecordNotFoundFailure("Request", id.toString())
```

Comment:
`RequestRepository#getById` already returns a failed future when request is not found:
```
.on(404, failed(new RecordNotFoundFailure(REQUESTS_COLLECTION_NAME, id))))
```